environment.py

```diff
@@ -49,42 +50,46 @@ def __init__(self, conf):
     
     def reset(self, options=None):
         ''' Choose initial state uniformly at random '''
-        self._state = np.zeros(self.conf.nb_state) 
+        state = np.zeros(self.conf.nb_state) 
 
-        rand_time = random.uniform(self.conf.x_init_min[-1], self.conf.x_init_max[-1])
+        time = random.uniform(self.conf.x_init_min[-1], self.conf.x_init_max[-1])
         for i in range(self.conf.nb_state-1): 
-            self._state[i] = random.uniform(self.conf.x_init_min[i], self.conf.x_init_max[i]) 
-        self._state[-1] = self.conf.dt*round(rand_time/self.conf.dt)
-        observation = self._state
+            state[i] = random.uniform(self.conf.x_init_min[i], self.conf.x_init_max[i]) 
+        state[-1] = self.conf.dt*round(time/self.conf.dt)
 
-        return rand_time, observation
+        return time, state
 
-    def step(self, rand_time, state, action):
+    def step(self, state, action):
         ''' Return next state and reward '''
         # compute next state
         state_next = self.simulate(state, action)
 
         # compute reward
-        reward = self.reward(rand_time, state_next, action)
+        reward = self.reward(state_next, action)
 
         return (state_next, reward)
+    
+    def simulate(self, state, action):
+        ''' Simulate dynamics '''
 
-    def get_end_effector_position(self, state, recompute=True):
-        ''' Compute end-effector position '''
-        nv = self.conf.robot.nv
         nq = self.conf.robot.nq
-        q = state[:nq] 
+        nv = self.conf.robot.nv
+        nx = nq + nv
 
-        RF = self.conf.robot.model.getFrameId(self.conf.end_effector_frame_id) 
+        state_next = np.zeros(nx+1)
 
-        H = self.conf.robot.framePlacement(q, RF, recompute)
-    
-        p = H.translation 
-        
-        return p
+        # Simulate control action
+        self.conf.simu.simulate(np.copy(state), action, self.conf.dt, 1)
+       
+        # Return next state
+        state_next[:nq], state_next[nq:nx] = np.copy(self.conf.simu.q), np.copy(self.conf.simu.v)
+        state_next[-1] = state[-1] + self.conf.dt
 
-    def reward(self, rand_time, state_next, action=None):
+        return state_next
+    
+    def reward(self, state_next, action=None):
```

Comment:
Do you think there's a way to avoid defining the reward twice? Both here and in the TO problem.